src/App/Application.cpp

```diff
@@ -452,46 +452,43 @@ void Application::renameDocument(const char *OldName, const char *NewName)
 
 Document* Application::newDocument(const char * Name, const char * UserName, bool createView, bool tempDoc)
 {
-    auto getNameAndLabel = [this](const char * Name, const char * UserName) -> std::tuple<std::string, std::string> {
-        bool defaultName = (!Name || Name[0] == '\0');
-
-        // get a valid name anyway!
-        if (defaultName) {
-            Name = "Unnamed";
-        }
+    std::string name;
+    bool defaultName = (!Name || Name[0] == '\0');
+    // get a valid name anyway!
+    if (defaultName) {
+        name = "Unnamed";
+    }
+    else {
+        name = Name;
+    }
 
-        std::string userName;
-        if (UserName && UserName[0] != '\0') {
-            userName = UserName;
-        }
-        else {
-            userName = defaultName ? QObject::tr("Unnamed").toStdString() : Name;
+    std::string userName;
+    if (UserName && UserName[0] != '\0') {
+        userName = UserName;
+    }
+    else {
+        userName = defaultName ? QObject::tr("Unnamed").toStdString() : Name;
 
-            std::vector<std::string> names;
-            names.reserve(DocMap.size());
+        if (!DocMap.empty()) {
+            // The assumption here is that there are not many documents and
+            // documents are rarely created so the cost
+            // of building this manager each time is inconsequential
+            Base::UniqueNameManager names;
             for (const auto& pos : DocMap) {
-                names.emplace_back(pos.second->Label.getValue());
+                names.addExactName(pos.second->Label.getValue());
             }
 
-            if (!names.empty()) {
-                userName = Base::Tools::getUniqueName(userName, names);
-            }
+            userName = names.makeUniqueName(userName);
         }
-
-        return std::make_tuple(std::string(Name), userName);
-    };
-
-    auto tuple = getNameAndLabel(Name, UserName);
-    std::string name = std::get<0>(tuple);
-    std::string userName = std::get<1>(tuple);
+    }
     name = getUniqueDocumentName(name.c_str(), tempDoc);

```

Comment:
Strange, I didn't find the lambda form at all easy to read...
I agree that some of the locals could have better names to make it clearer what they actually mean. In particular the name/userName is confusing when everywhere else calls them name and label.
Though I would use the "proposed" prefixes for the arguments, these are the values proposed by the caller.
Some of the complexity comes from the desire to translate the default label but not the default name, leading to the need for the clumsy defaultName boolean. Again renaming this helps.

Reply:
Actually, I'm tempted to roll getUniqueDocumentName into newDocument since this is the only real call to it.
There are about 20 calls to it in the test suite, but they all them immediately call newDocument passing the name they just got.